[libc++] Check iterator ownership in erase() for string and vector

cjdb opened the pull request:
This ensures that the iterators passed in are in-range for the container. Thanks to @mclow for identifying this opportunity.

cjdb committed:
[libcxx] adds additional checks to RAI containers' `erase`

This ensures that the iterators passed in are in-range for the
container. Thanks to @mclow for identifying this opportunity.

cjdb commented:
I haven't added tests yet, since I want to check that this is the direction we intend to head.

cjdb commented:
Ping

ldionne reviewed (commented):
Thanks for implementing this improvement!

Can you please add tests for those new assertions?

cjdb committed:
Update libcxx/include/deque

Co-authored-by: Louis Dionne <ldionne.2@gmail.com>

ldionne committed:
- Use __is_pointer_in_range for a non-UB comparison
- Add tests
- Downgrade checks we do in deque cause we can't detect iterator ownership for non-contiguous containers

ldionne committed:
Formatting

ldionne commented:
This is slipping past the LLVM 19 release.
